### ISSUE TITLE
stop codon mutations

### ISSUE THREAD

**Issue by reimand0** [reimandlab/ActiveDriverDB#94]
how does the current DNA-to-protein mutation conversion handle and report stop-codon mutations, such as those that cause a new stop codon in the middle of the protein, or those that remove the natural stop codon at the end of the protein?

**Comment by krassowski** [comment]
I have just checked this for the first scenario:

https://rl-db.oicr.on.ca/chromosome/mutation/chr17/19284200/G/C has tryptophan as reference aminoacid, which is encoded only by UGG codon. Changing the second G to C changes aminoacid to cysteine and changing base on this position to A makes it a stop codon:
https://rl-db.oicr.on.ca/chromosome/mutation/chr17/19284200/G/A
**What returns an empty response.**

Running grep on the input data for DNA-protein mapping confirms that such mutations are not enlisted in the data:

```bash
zgrep -ce ":c\.G[0-9]*C:p\.W" annot_*.txt.gz
```

**Comment by krassowski** [comment]
We can change this behaviour with code like:
```
if G→A gives an empty response:
      if G→C in the same position has ref == "W" and alt == "C", then:
           change response to "stop codon gained" 
```
For SNVs which cause "gain of stop codon" events we will need to consider:
- 3 + 2 + 3 SNVs for TGA stop codon,
- 3 + 2 + 2 for TAA stop codon,
- 3 + 3 + 2 for TAG stop codon

I can hardcode this behaviour and it shouldn't cause a big performance issues. In worst scenario those are several queries more to the mappings database to be done.

I think that the same applies to removals of stop codons.

There are also "synonymous" mutations in stop codons - should we also consider those?

**Comment by reimand0** [comment]
Stop codons are important to annotate because these change the resulting
protein. It would be great if you could add this functionality. Silent
(synonymous) mutations are ok to ignore. In the protein sense they are
equivalent to mutations outside of the genes.

**Comment by krassowski** [comment]
I am sorry, but it seems that I was wrong about the "stop codon lost" events.

I successfully implemented detection of "stop gained" events because the information about codons underlying specific aminoacids can be easily derived from the SNV → CSV.

Conversely, I cannot determine if a stop codon was destroyed, using only the data that we already have in the database - simply there is no information about codons which were present in DNA sequence on specific positions.

To be more precise: I can check if there is a stop on given position (looking up isoforms table for an entry where tx_end is close to this position), I can tell that some mutations (any base→C, first base → anything) destroy the stop codon. The problem is with the second and third base of the codon: I need to know which codon do I have, to distinguish between the "silent" and "stop lost" effects of mutations.

Basically it seems another approach would be needed. I think about precomputing this data or importing DNA sequences for proteins we have. Do you have any suggestions?

**Comment by reimand0** [comment]
let's just ignore these mutations and add a small note in the documentation that stop codon mutations are not considered. please close when ready.
